fendermint/vm/topdown/src/voting.rs

```diff
@@ -0,0 +1,249 @@
+        let (height, hash) = last_finalized_block;
+        Self {
+            power_table: TVar::new(im::HashMap::from_iter(power_table)),
+            chain: TVar::new(im::OrdMap::from_iter([(height, Some(hash))])),
+            votes: TVar::default(),
+            pause_votes: TVar::new(false),
+        }
+    }
+
+    /// Check that a validator key is currently part of the power table.
+    pub fn known_validator(&self, validator_key: &K) -> Stm<bool> {
+        let pt = self.power_table.read()?;
+        // For consistency consider validators without power unknown.
+        match pt.get(validator_key) {
+            None => Ok(false),
+            Some(weight) => Ok(*weight > 0),
+        }
+    }
+
+    /// Calculate the minimum weight needed for a proposal to pass with the current membership.
+    pub fn quorum_threshold(&self) -> Stm<Weight> {
+        let total_weight: Weight = self.power_table.read().map(|pt| pt.values().sum())?;
+
+        Ok(total_weight * 2 / 3)
+    }
+
+    /// Return the height of the first entry in the chain.
+    pub fn last_finalized_height(&self) -> Stm<BlockHeight> {
+        self.chain
+            .read()
+            .map(|c| c.get_min().map(|(h, _)| *h).unwrap_or_default())
+    }
+
+    /// Add the next final block observed on the parent blockchain.
+    ///
+    /// Returns an error unless it's exactly the next expected height.
+    pub fn add_block(
+        &self,
+        block_height: BlockHeight,
+        block_hash: Option<V>,
+    ) -> StmResult<(), Error<K>> {
+        let mut chain = self.chain.read_clone()?;
+
+        // Check that we are extending the chain. We could also ignore existing heights.
+        if let Some((prev, _)) = chain.get_max() {
+            if block_height != prev + 1 {
+                return abort(Error::UnexpectedBlock(prev + 1, block_height));
+            }
+        }
+
+        chain.insert(block_height, block_hash);
+
+        self.chain.write(chain)?;
+
+        Ok(())
+    }
+
+    /// Add a vote we received.
+    ///
+    /// Returns `true` if this vote was added, `false` if it was ignored as a
+    /// duplicate or a height we already finalized, and an error if it's an
+    /// equivocation or from a validator we don't know.
+    pub fn add_vote(
+        &self,
+        validator_key: K,
+        block_height: BlockHeight,
+        block_hash: V,
+    ) -> StmResult<bool, Error<K, V>> {
+        let min_height = self.last_finalized_height()?;
+
+        if block_height < min_height {
+            return Ok(false);
+        }
+
+        if !self.known_validator(&validator_key)? {
+            return abort(Error::UnknownValidator(validator_key));
+        }
+
+        let mut votes = self.votes.read_clone()?;
+        let votes_at_height = votes.entry(block_height).or_default();
+
+        for (bh, vs) in votes_at_height.iter() {
+            if *bh != block_hash && vs.contains(&validator_key) {
```

Comment:
ah, so `Equivocation` is for when a validator votes on a height with different hash than he did previously?